core/direct-rpc-server/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+
+*/
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(test, feature(assert_matches))]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use jsonrpc_core_sgx as jsonrpc_core;
+	pub use thiserror_sgx as thiserror;
+}
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{rpc_connection_registry::ConnectionRegistry, rpc_watch_extractor::RpcWatchExtractor};
+use codec::{Encode, Error as CodecError};
+use itc_tls_websocket_server::{WebSocketConnection, WebSocketError};
+use itp_types::{RpcResponse, TrustedOperationStatus};
+use serde_json::error::Error as SerdeJsonError;
+use sp_runtime::traits;
+use std::{fmt::Debug, vec::Vec};
+
+#[cfg(test)]
+mod mocks;
+
+#[cfg(test)]
+mod builders;
+
+pub mod rpc_connection_registry;
+pub mod rpc_responder;
+mod rpc_watch_extractor;
+pub mod rpc_ws_handler;
+
+/// General web-socket error type
+#[derive(Debug, thiserror::Error)]
+pub enum DirectRpcError {
+	#[error("Invalid connection hash")]
+	InvalidConnectionHash,
+	#[error("RPC serialization error: {0}")]
+	SerializationError(SerdeJsonError),
+	#[error("Web socket error: {0}")]
+	WebSocketError(WebSocketError),
+	#[error("Encoding error: {0}")]
+	EncodingError(CodecError),
+}
+
+pub type DirectRpcResult<T> = Result<T, DirectRpcError>;
+
+/// trait helper to mix-in all necessary traits for a hash
+pub trait RpcHash: std::hash::Hash + traits::Member + Encode {}
+impl<T: std::hash::Hash + traits::Member + Encode> RpcHash for T {}
+
+/// registry for RPC connections (i.e. connections that are kept alive to send updates)
+pub trait RpcConnectionRegistry: Send + Sync {
+	type Hash: RpcHash;
+	type Connection: WebSocketConnection;
+
+	fn store(&self, hash: Self::Hash, connection: Self::Connection, rpc_response: RpcResponse);
+
+	fn withdraw(&self, hash: &Self::Hash) -> Option<(Self::Connection, RpcResponse)>;
+}
+
+/// sends an RPC response back to the client
+pub trait SendRpcResponse: Send + Sync {
+	type Hash: RpcHash;
+
+	fn update_status_event(
+		&self,
+		hash: Self::Hash,
+		status_update: TrustedOperationStatus,
+	) -> DirectRpcResult<()>;
+
+	fn send_state(&self, hash: Self::Hash, state_encoded: Vec<u8>) -> DirectRpcResult<()>;
+}
+
+/// determines if a given connection must be watched (i.e. kept alive),
+/// based on the information in the RpcResponse
+pub trait DetermineWatch {
+	type Hash: RpcHash;
+
+	fn must_be_watched(&self, rpc_response: &RpcResponse) -> DirectRpcResult<Option<Self::Hash>>;
+}
+
+/// convenience method to create a connection registry
+pub fn create_connection_registry<Hash, Connection>() -> ConnectionRegistry<Hash, Connection>
```

Comment:
The idea was to keep the modules in that crate private and have all the public API in `lib.rs`. (I did not succeed completely though, had to make some modules public again)

Reply:
Ok, I see. Just be aware that we try to be an sdk. Even if we don't need some modules to be public, we should think of potentially other user's needs..

Reply:
It's certainly good to keep that in mind - and I think a clean encapsulation inside the crates helps with that goal to clearly state which parts are crate private and which parts make up the public API of a crate.

Reply:
This is something I just realized, I don't know if that is a problem. Just for the future, you have here a wrapper that returns an `RPCWatchExtractor`. But the module that contains it is private. So at least `RPCWatchExtractor` should be made public. Otherwise, you can't import the struct in case the compiler needs type help with type inference.